### PR TITLE
DOC: Add doc-string examples for pd.read_sql using custom parse_dates arg values

### PR DESCRIPTION
**Pull request by avinashpancham**
follow on PR  for  #37823

- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/sql.py [hunk 1]
```diff
@@ -478,6 +478,41 @@ def read_sql(
     -------
     DataFrame or Iterator[DataFrame]
 
+    Examples
+    --------
+    Read data from SQL via either a SQL tablename or a SQL query
+
+    >>> pd.read_sql('table_name', 'postgres:///db_name')  # doctest:+SKIP
```
Comment (mroeschke) [inline pandas/io/sql.py:485]: Maybe the examples can demonstrate connecting to a sqlite db so we don't have to skip all these doctests

File: pandas/io/sql.py [hunk 2]
```diff
@@ -478,6 +478,41 @@ def read_sql(
...
+
+    >>> pd.read_sql('SELECT * FROM table_name', 'postgres:///db_name')  # doctest:+SKIP
+
+    Apply dateparsing to columns through the "parse_dates" argument
+
+    >>> pd.read_sql('table_name',
+    ...             'postgres:///db_name',
+    ...             parse_dates=["date_column"])  # doctest:+SKIP
+
+    The "parse_dates" argument calls pd.to_datetime on the provided columns. Custom
```
Comment (mroeschke) [inline pandas/io/sql.py:495]: Double backticks around ``pd.to_datetime``

File: pandas/io/sql.py [hunk 3]
```diff
@@ -478,6 +478,41 @@ def read_sql(
...
 
+    Examples
+    --------
+    Read data from SQL via either a SQL tablename or a SQL query
+
+    >>> pd.read_sql('table_name', 'postgres:///db_name')  # doctest:+SKIP
+
+    >>> pd.read_sql('SELECT * FROM table_name', 'postgres:///db_name')  # doctest:+SKIP
+
+    Apply dateparsing to columns through the "parse_dates" argument
```
Comment (mroeschke) [inline pandas/io/sql.py:489]: Double backticks around ``parse_dates`` instead

File: pandas/io/sql.py (hunk 1 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 2 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 3 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

**Comment by avinashpancham** [comment]
Added doctests pass. 

CI fails due to other reasons, since I only added documentation. Will merge master later to run CI agian.

File: pandas/io/sql.py [hunk 4]
```diff
@@ -482,6 +482,62 @@ def read_sql(
     --------
     read_sql_table : Read SQL database table into a DataFrame.
     read_sql_query : Read SQL query into a DataFrame.
+
+    Examples
+    --------
+    Read data from SQL via either a SQL query or a SQL tablename (latter not
+    possible for SQLite tables)
+    >>> from sqlite3 import connect
+    >>> conn = connect('file.db')
```
Comment (mroeschke) [inline pandas/io/sql.py:491]: You can use `":memory:"` instead

File: pandas/io/sql.py [hunk 5]
```diff
@@ -482,6 +482,62 @@ def read_sql(
     --------
     read_sql_table : Read SQL database table into a DataFrame.
     read_sql_query : Read SQL query into a DataFrame.
+
+    Examples
+    --------
+    Read data from SQL via either a SQL query or a SQL tablename (latter not
+    possible for SQLite tables)
```
Comment (mroeschke) [inline pandas/io/sql.py:489]: Can you add a newline here

File: pandas/io/sql.py [hunk 6]
```diff
@@ -482,6 +482,62 @@ def read_sql(
...
+    >>> df.to_sql('test_data', conn)
+
+    >>> pd.read_sql('SELECT int_column, date_column FROM test_data', conn)
+       int_column date_column
+    0           0    10/11/12
+    1           1    12/11/10
+
+    >>> pd.read_sql('test_data', 'postgres:///db_name')  # doctest:+SKIP
+
+    Apply dateparsing to columns through the ``parse_dates`` argument
```
Comment (mroeschke) [inline pandas/io/sql.py:503]: `date parsing`

File: pandas/io/sql.py [hunk 7]
```diff
@@ -482,6 +482,62 @@ def read_sql(
...
+
+    >>> pd.read_sql('SELECT int_column, date_column FROM test_data',
+    ...             conn,
+    ...             parse_dates=["date_column"])
+           int_column date_column
```
Comment (mroeschke) [inline pandas/io/sql.py:508]: These column names looks offset

File: pandas/io/sql.py [hunk 8]
```diff
@@ -482,6 +482,62 @@ def read_sql(
...
+
+    The ``parse_dates`` argument calls ``pd.to_datetime`` on the provided columns.
+    Custom argument values for applying ``pd.to_datetime`` on a column are specified
+    via a dictionary format:
+    1. Ignore errors while parsing the values of "date_column"
+
+    >>> pd.read_sql('SELECT int_column, date_column FROM test_data',
+    ...             conn,
+    ...             parse_dates={"date_column": {"errors": "ignore"}})
+           int_column date_column
```
Comment (mroeschke) [inline pandas/io/sql.py:520]: Same

File: pandas/io/sql.py [hunk 9]
```diff
@@ -482,6 +482,62 @@ def read_sql(
...
+       int_column date_column
+    0           0  2012-11-10
+    1           1  2010-11-12
+
+    3. Apply custom formatting when dateparsing the values of "date_column"
+
+    >>> pd.read_sql('SELECT int_column, date_column FROM test_data',
+    ...             conn,
+    ...             parse_dates={"date_column": {"format": "%d/%m/%y"}})
+           int_column date_column
```
Comment (mroeschke) [inline pandas/io/sql.py:538]: Same

File: pandas/io/sql.py (hunk 5 above)
Comment (mroeschke) [inline pandas/io/sql.py:489]: I would assume `parse_dates` should still work for sqlite. 

Our docs state

```
Dict of {column_name: arg dict}, where the arg dict corresponds to the keyword arguments of pandas.to_datetime() Especially useful with databases without native Datetime support, such as SQLite.
```

Comment (avinashpancham) [reply pandas/io/sql.py:489]: Sorry, I will clarify this. What I mean is that SQLite only accepts queries not just table names:
```python
pd.read_sql('SELECT * FROM test_data', conn) # works for SQLite
pd.read_sql('test_data', conn) # does not work for SQLite
```

File: pandas/io/sql.py (hunk 4 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Cool, didn't know this! Changed it.

File: pandas/io/sql.py (hunk 5 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 6 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 7 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 8 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 9 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: pandas/io/sql.py (hunk 5 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Made explanation more explicit

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
very nice @avinashpancham

**Pull request merged**